File: src/project_euler/problem_6.clj
(ns project-euler.problem-6)

(defn solver
  "Returns the difference between sum-of-squares and square-of-sum of first n integers."
  [n]
  (letfn [(sum-of-squares [n]
                          (* (/ 1 6)
                             n
                             (+ n 1)
                             (+ (* 2 n) 1)))
          (square-of-sum [n]
                         (let [sum (* (/ 1 2)
                                      n
                                      (+ n 1))]
                           (* sum sum)))]
    (- (square-of-sum n)
       (sum-of-squares n))))


(def problem
  "Find the difference between the sum of the squares of the first one hundred natural numbers and the square of the sum.")

(def solution
  (solver 100))

File: src/problem6/Problem6.java
/**
 *	@author		Arshan Alam
 *
 **/

package problem6;
/**
 * 	<p>
 *		Problem 6
 *	</p>
 *
 *	<p>
 *		The sum of the squares of the first ten natural numbers is,
 *	</p>
 *
 *	<p>
 *		1^2 + 2^2 + ... + 10^2 = 385
 *	</p>
 *
 *	<p>
 *		The square of the sum of the first ten natural numbers is,
 *	</p>
 *
 *	<p>
 *		(1 + 2 + ... + 10)^2 = 55^2 = 3025
 *	</p>
 *
 *	<p>
 *		Hence the difference between the sum of the squares of the
 *		first ten natural numbers and the square of the sum is 3025 − 385 = 2640.
 *	</p>
 *
 *	<p>
 *		Find the difference between the sum of the squares of
 *		the first one hundred natural numbers and the square of the sum.
 *	</p> 
 **/
public class Problem6 {
	public static void main(String[] args) {
		/**
		 * Solution uses facts from Mathematics
		 *
		 * sum[i: 1..n](i) = (n)(n+1)/2
		 *
		 * sum[i: 1..n](i^2) = (n)(n+1)(2n+1)/6
		 **/
		System.out.println(
				(((100*(100+1))/2)*((100*(100+1))/2))
				-
				((100*(100+1)*(2*100 + 1))/6));
	}	// end of main
	
}	// end of class Problem6